lib/exceptional_synchrony/event_machine_proxy.rb

```diff
@@ -79,16 +83,23 @@ def run(on_error: :log, &block)
       end
     end
 
-    def defer(context, &block)
-      deferrable = EventMachine::DefaultDeferrable.new
-
-      callback = -> (result) { deferrable.succeed(result) }
-
-      EventMachine.defer(nil, callback) { CallbackExceptions.return_exception(&block) }
+    # This method will execute the block on the background thread pool
+    # By default, it will block the caller until the background thread has finished, so that the result can be returned
+    #  :wait_for_result - setting this to false will prevent the caller from being blocked by this deferred work
+    def defer(context, wait_for_result: true, &block)
+      if wait_for_result
+        deferrable = EventMachine::DefaultDeferrable.new
+        callback = -> (result) { deferrable.succeed(result) }
 
-      EventMachine::Synchrony.sync(deferrable)
-
-      CallbackExceptions.map_deferred_result(deferrable)
+        EventMachine.defer(nil, callback) { CallbackExceptions.return_exception(&block) }
+        EventMachine::Synchrony.sync(deferrable)
+        CallbackExceptions.map_deferred_result(deferrable)
+      else
+        EventMachine.defer do
```

Comment:
This might be more symmetrical if on one line? (Or, leave it multi-line and change the above to match.) Also, it can pass `&block`:
```
EventMachine.defer { ExceptionHandling.ensure_completely_safe("defer", &block) }

Reply:
I like that, updated